flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -635,6 +641,33 @@ public void cancel() {
         return executionGraph.getTerminationFuture().thenApply(FunctionUtils.nullFn());
     }
 
+    protected void archiveGlobalFailure(Throwable failure) {
+        taskFailureHistory.add(
+                new ErrorInfo(failure, executionGraph.getStatusTimestamp(JobStatus.FAILED)));
+        log.debug("Archive global failure.", failure);
+    }
+
+    protected void archiveFromFailureHandlingResult(FailureHandlingResult failureHandlingResult) {
+        Optional<Execution> executionOptional =
+                failureHandlingResult
+                        .getExecutionVertexIdOfFailedTask()
+                        .map(this::getExecutionVertex)
+                        .map(ExecutionVertex::getCurrentExecutionAttempt);
+
+        executionOptional.ifPresent(
+                execution ->
+                        execution
+                                .getFailureInfo()
+                                .ifPresent(
+                                        failureInfo -> {
+                                            taskFailureHistory.add(failureInfo);
+                                            log.debug(
+                                                    "Archive local failure causing attempt {} to fail: {}",
+                                                    execution.getAttemptId(),
+                                                    failureInfo.getExceptionAsString());
+                                        }));
```

Comment:
Why do we have to ask the execution for the failure cause? If we are only interested in the root cause, doesn't `failureHandlingResult` contain all the required information?

Reply:
Because of the timestamp in case of local failure: The `Execution` fails but does not necessarily trigger the `ExecutionGraph` to switch to fail. Hence, no timestamp is record for this kind of failure.

Reply:
But looking into that again, I realize that there is, indeed, the `failureInfo` in `ExecutionGraph` that is set during [handleTaskFailure](https://github.com/XComp/flink/blob/8e732bfb2bddc38ec7422f482dcda4be3d296408/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java#L218) and [handleGlobalFailure](https://github.com/XComp/flink/blob/8e732bfb2bddc38ec7422f482dcda4be3d296408/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java#L236). This is used to forward the current root cause of the `ExecutionGraph` to the Web UI. I could have used that. But the drawback is that it's using `System.currentMillis`.

Instead, I'm gonna work on removing the `ExecutionGraph.failureInfo` (and related methods) as part of FLINK-21188. It becomes redundant after we enable to exception history since it's just the exception history's last entry.

Reply:
So theoretically we would have to add the timestamp to the `FailureHandlingResult` or the `ErrorInfo` if we didn't want to look up the `Execution` again, right?

Reply:
Yes, that would be possible if would would like to relax the constraint that the exception history does not provide the timestamp the `Execution` was marked as failed but an earlier timestamp. But to be fair, we do that already for the fallback of a global failover in `archiveFromFailureHandlingResult`.

Reply:
Ok, I think we don't have to change it right now. Maybe we do this sometime in the future.

Reply:
I added this remark to FLINK-21190